app/src/helpers/helpers.ts

```diff
@@ -59,8 +59,15 @@ export function linkIsInternal(
     // 1. app.foo.com and foo.com
     // 2. www.foo.com and foo.com
     // 3. www.foo.com and app.foo.com
-    const currentDomain = new URL(currentUrl).hostname.replace(/^www\./, '');
-    const newDomain = new URL(newUrl).hostname.replace(/^www./, '');
+
+    // Only use the tld and the main domain for domain-ish test
+    // Enables domain-ish equality for blog.foo.com and shop.foo.com
+    // Also makes the endWith check basically obsolete, === could be used.
+    const currentDomain = new URL(currentUrl).hostname
+      .split('.')
+      .slice(-2)
```

Comment:
This logic will break on [Second-Level Domains](https://en.wikipedia.org/wiki/Second-level_domain) like `.co.uk`. Try adding a similar test to the one you added, but instead of

```
const sameBaseDomainUrlTidalListen = 'https://listen.tidal.com/';
const sameBaseDomainUrlTidalLogin = 'https://login.tidal.com/';
```
, you work on

```
const sameBaseDomainUrlUkTidalListen = 'https://listen.tidal.co.uk/';
const sameBaseDomainUrlUkTidalLogin = 'https://login.tidal.co.uk/';
```

There is no "clean" solution to this. The only way to do it is to use a lib that packages the list of SLDs, but I want to avoid this, because it's ever changing. Thus the current logic with a simple special case for `www.`

Anyway, I think your problem is better solved by making `isInternalLoginPage` return true for non-specific domains matching `(login|auth|account|accounts|signin).[a-zA-Z\\.]*` . Objections to this, @TheCleric ?

Reply:
Thanks for the review @ronjouch 

Tried to think of a better solution, but without any list of TLDs, this is basically not possible.

I personally prefer the second solution, but the regex should be
```
(login|auth|account|accounts|signin)\\.[a-zA-Z0-9_.-]+

a) matches the dot after (login|auth|account|accounts|signin) correctly
b) the - in the character class has to be at the end or escaped if I'm not wrong
```
to match something like `login.tidal.com`, shouldn't it?

Reply:
@rippin93 thanks, yeah I think you're right about needing to put the `-` at the end of the regex. Also, maybe we need something better than `a-zA-Z` to cover for unicode domains? Or are they punycode at this point? I'll fiddle a bit in my REPL when working on this.

Waiting for a 👍 from @TheCleric , and will do this. In the meantime, you can work around the issue with a `--internal-urls`

Reply:
Well this is very good, validation of unicode strings or utf-8 encoded string can be very annoying.

Will work on this probably tomorrow (meetings can take time :unamused:), think we should give @TheCleric a bit time to review and answer to this :+1:

Reply:
I think I have an even simpler solution that I have tried against @rippin93 's branch and all tests pass even with @ronjouch 's new tests.

I don't like pushing changes to other people's repos that are more than simple fixes, so I'll show my work here.

```typescript
export function linkIsInternal(
  currentUrl: string,
  newUrl: string,
  internalUrlRegex: string | RegExp,
): boolean {
  if (newUrl === 'about:blank') {
    return true;
  }

  if (isInternalLoginPage(newUrl)) {
    return true;
  }

  if (internalUrlRegex) {
    const regex = RegExp(internalUrlRegex);
    if (regex.test(newUrl)) {
      return true;
    }
  }

  try {
    // Consider as "same domain-ish", without TLD/SLD list:
    // 1. app.foo.com and foo.com
    // 2. www.foo.com and foo.com
    // 3. www.foo.com and app.foo.com

    // Only use the tld and the main domain for domain-ish test
    // Enables domain-ish equality for blog.foo.com and shop.foo.com
    return domainify(currentUrl) === domainify(newUrl);
  } catch (err) {
    console.warn(
      'Failed to parse domains as determining if link is internal. From:',
      currentUrl,
      'To:',
      newUrl,
      err,
    );
    return false;
  }
}

function domainify(url: string): string {
  // So here's what we're doing here:
  // Get the hostname from the url
  const hostname = new URL(url).hostname;
  // Drop the first section if the domain
  const domain = hostname.split('.').slice(1).join('.');
  // Check the length, if it's too short, the hostname was probably the domain
  // Or if the domain doesn't have a . in it we went too far
  if (domain.length < 6 || domain.split('.').length === 0) {
    return hostname;
  }
  // This SHOULD be the domain, but nothing is 100% guaranteed
  return domain;
}
```

Reply:
@TheCleric 👍 to your new attempt, feels like it covers a bit more ground indeed, even if as you wrote it's not 100% reliable (but still better than what we have). Open to it. BTW, I was thinking we could also have tests for:

1. Internal sites (e.g. http://localhost , http://othermachine:8080). But I'm not sure what we want to assert, in these cases. Should http://roundcube.localhost and http://gogs.localhost be considered internal? Iiiii don't think so, since typically they're different apps?)
2. Multi-nested sites (e.g. https://terrible.enterprise.app.emea.sales.companyname.com and https://otherapp.emea.sales.companyname.com ). Same question here: whaaaat the hell do we want to do? I'd wager to consider them external, for the same reason?!

Thoughts?

Reply:
> @TheCleric 👍 to your new attempt, feels like it covers a bit more ground indeed, even if as you wrote it's not 100% reliable (but still better than what we have). Open to it. BTW, I was thinking we could also have tests for:
> 
> 1. Internal sites (e.g. http://localhost , http://othermachine:8080). But I'm not sure what we want to assert, in these cases. Should http://roundcube.localhost and http://gogs.localhost be considered internal? Iiiii don't think so, since typically they're different apps?)
> 2. Multi-nested sites (e.g. https://terrible.enterprise.app.emea.sales.companyname.com and https://otherapp.emea.sales.companyname.com ). Same question here: whaaaat the hell do we want to do? I'd wager to consider them external, for the same reason?!
> 
> Thoughts?

I think at this point we wait until someone has a need for that to worry about it. 😆

Reply:
Hey @TheCleric @ronjouch,

this looks pretty good, @TheCleric I'll put your proposition in my PR (and I'll also put a little `new` before the `Regexp(internalUrlRegex)` if you don't mind ;)

That being said, I think I'll play around a little more with this, I think there could be even more potential.

Reply:
Found an interesting library: https://github.com/medialize/URI.js

Any objections just using this to determine the domain?

Reply:
I also implemented a version with the package [tldts](https://www.npmjs.com/package/tldts).

Would be happy if you could take a look at this, I guess it could be a valuable addition here :grin:

Reply:
> I also implemented a version with the package [tldts](https://www.npmjs.com/package/tldts).
> 
> 
> 
> Would be happy if you could take a look at this, I guess it could be a valuable addition here :grin: 
> 
> 
> 
> If you don't like it, I'll just remove it again :man_shrugging: 

I think we had previously been using that library (or something similar) here, but had removed it to keep out extra dependencies, especially in the final generated app.

Reply:
> I also implemented a version with the package [tldts](https://www.npmjs.com/package/tldts). Would be happy if you could take a look at this, I guess it could be a valuable addition here. If you don't like it, I'll just remove it again

@rippin93 thanks

> I think we had previously been using that library (or something similar) here, but had removed it to keep out extra dependencies, especially in the final generated app.

@TheCleric correct, see 6b266b78150ca8617ba248cca19325d003524ac6  and c0a660467665009281f9066c9f426aeb9cb9f87b , in which I wrote (slightly altering
the second paragraph to merge what we did in the two commits):

> **(Attempt to) get rid of deprecated app dep wurl**
>
> This one may be problematic, as it used to do TLD stuff:
> https://github.com/websanova/node-url/blob/7982a613bc/wurl.js#L4
>
> So, the new WHATWG-URL-based implementation will consider
> `shop.asana.com` to be "external" to `app.asana.com`, contrarily to before.
> Given the nature of Nativefier, I think it's actually what to expect,
> that in this case one is "out of the app", and in e.g. asana landing's page,
> which you'd expect to see in your browser.
>
> Let's see if users disagree with that.

So, @rippin93 you found a login use case that hits this road bump, but apart from you, I saw no reports about the "blog subdomain" <-> "app subdomain", so:

1. It feels like the assumption I made in this commit wasn't entirely bad (as we only got one user bug since)
2. I maintain I'd like to cover your case anyway with our `isInternalLoginPage` logic
3. I maintain that I like a **lot** to stay as nimble as possible with dependencies (I wrote in our [dev guide](https://github.com/nativefier/nativefier/blob/master/docs/development.md#guidelines): _"Avoid or limit adding npm dependencies. Each new dependency is a complexity & security liability."_). That being said, @rippin93, it's true that the library `tldts` you found seems like a good one (no dependencies, focus on perf, etc).

→ Given all these,

1. I'd like to give a last try to the last "best effort" dependency-less and list-of-SLD-less solution offered by @TheCleric .
2. I'd like to go with the `isInternalLoginPage` generic-ification / described at https://github.com/nativefier/nativefier/pull/1171#discussion_r623813960

Makes sense / am I omitting to consider something?

Reply:
Well for me, any solution is fine, as long as it doesn‘t come with a nasty part like my first attempt 😉 

God bless git which allows me to just reset to my last commit, but it‘s already sleeping time for me today, so I tend to change this tomorrow 👍 

Will also take into Account the change of the Login regexes you mentioned in your comment.

Have a good night guys 😴

Reply:
@rippin93 👍. BTW,

- Rather than undo your last commit, `git revert` it! That way, we'll keep your `tldts` experiment in the PR branch, ready for use if we end up wanting it.
- Also, I expanded in [Dev guide / guidelines](https://github.com/nativefier/nativefier/blob/master/docs/development.md#guidelines) on the reasons for me to be such a party pooper with deps. Copy-pasta:
    > **Avoid adding npm dependencies**. Each new dep is a complexity & security liability. You might be thinking your extra dep is _"just a little extra dep"_, and maybe you found one that is high-quality & dependency-less. Still, it's an extra dep, and over the life of Nativefier we requested changes to _dozens_ of PRs to avoid "just a little extra dep". Without this constant attention, Nativefier would be more bloated, less stable for users, more annoying to maintainers. Now, don't go rewriting zlib if you need a zlib dep, for sure use a dep. But if you can write a little helper function saving us a dep for a mundane task, go for the helper :) . Also, an in-tree helper will always be less complex than a dep, as inherently more tailored to our use case, and less complexity is good.

Reply:
@ronjouch alright, will do this in a free minute :+1: